From 9eefa3b4159a33f94d5ff8b277da9e6dcca6a897 Mon Sep 17 00:00:00 2001
From: Matthias Clasen <mclasen@redhat.com>
Date: Wed, 15 Oct 2025 18:21:30 -0400
Subject: [PATCH] Fix scaling of svg images

When generating more pixels than the width x height ask for,
we need to wrap the texture in a GtkScaler to preserve the
scaling information.

Closes #7823

Part-of: <https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/9045>
---
 gtk/gdktextureutils.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/gtk/gdktextureutils.c b/gtk/gdktextureutils.c
index f885a5f30a0..5d02b1718f4 100644
--- a/gtk/gdktextureutils.c
+++ b/gtk/gdktextureutils.c
@@ -703,10 +703,21 @@ gdk_paintable_new_from_bytes_scaled (GBytes *bytes,
       /* We know these formats can't be scaled */
       return GDK_PAINTABLE (gdk_texture_new_from_bytes (bytes, NULL));
     }
-  else
+  else if (scale == 1)
     {
       return GDK_PAINTABLE (gdk_texture_new_from_svg_bytes (bytes, scale, NULL));
     }
+  else
+    {
+      GdkTexture *texture;
+      GdkPaintable *paintable;
+
+      texture = gdk_texture_new_from_svg_bytes (bytes, scale, NULL);
+      paintable = gtk_scaler_new (GDK_PAINTABLE (texture), scale);
+      g_object_unref (texture);
+
+      return paintable;
+    }
 }
 
 GdkPaintable *
-- 
GitLab

